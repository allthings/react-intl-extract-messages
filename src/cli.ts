#!/usr/bin/env node
import fs from 'fs'
import path from 'path'

import minimist from 'minimist'

const argv = minimist(process.argv.slice(2))
const extractMessages = require('./index')

if (argv.help || !argv.tsconfig) {
  console.log(
    'Usage: message-extractor --tsconfig tsconfig.json [--out file.json]'
  )
  process.exit(argv.tsconfig ? 0 : 1)
}

const tsconfigFilePath = path.join(process.cwd(), argv.tsconfig)

const messages = JSON.stringify(extractMessages(tsconfigFilePath), null, 2)

if (!argv.out) {
  console.log(messages)
} else {
  fs.writeFileSync(path.join(process.cwd(), argv.out), messages, 'utf-8')
}
