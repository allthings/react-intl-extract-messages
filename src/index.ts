import childProcess from 'child_process'
import fs from 'fs'
import path from 'path'

import glob from 'glob'
import tmp from 'tmp'
import {
  transformFileSync,
  BabelFileMetadata,
  BabelFileResult,
} from '@babel/core'

const projectDir = path.dirname(process.argv[1])

interface IntlMessage {
  id: string
  defaultMessage: string
  description?: string
}

interface BabelMetadataReactIntl extends BabelFileMetadata {
  'react-intl': {
    messages: IntlMessage[]
  }
}

interface BabelFileResultReactIntl extends BabelFileResult {
  metadata: BabelMetadataReactIntl
}

export function extractMessages(tsconfigFile: string) {
  const tsconfigFilePath = path.isAbsolute(tsconfigFile)
    ? tsconfigFile
    : path.join(projectDir, tsconfigFile)

  if (!fs.existsSync(tsconfigFilePath)) {
    throw new Error('Provide a valid tsconfig file.')
  }

  // `unsafeCleanup` removes the folder recursively
  const tmpDir = tmp.dirSync({ unsafeCleanup: true })

  const typescriptBinary = path.join(
    // tsconfig dir indicates the root of the project
    path.dirname(tsconfigFilePath),
    'node_modules/.bin/tsc',
  )
  const args = [
    `--project ${tsconfigFilePath}`,
    `--outDir ${tmpDir.name}`,
    // babel-plugin-react-intl relies on imports
    `--module esnext`,
    `--jsx preserve`,
    `--sourceMap false`,
    '--declaration false',
  ].join(' ')

  try {
    // needs a spawned shell, so execFileSync doesn't work here
    childProcess.execSync(`${typescriptBinary} ${args}`)
  } catch (e) {
    tmpDir.removeCallback()
    throw new Error(e.stdout.toString())
  }

  const files = glob.sync(path.join(tmpDir.name, '**/*.@(js|jsx)'))

  const babelCompileOptions = {
    presets: ['@babel/preset-react'],
    plugins: [
      require('@babel/plugin-syntax-dynamic-import'),
      require('babel-plugin-react-intl'),
    ],
  }
  const messages: IntlMessage[] = files.reduce(
    (allMessages: IntlMessage[], fileName) => {
      const result = transformFileSync(
        fileName,
        babelCompileOptions,
      ) as BabelFileResultReactIntl

      const { messages } = result.metadata['react-intl']

      return [...allMessages, ...messages]
    },
    [],
  )

  tmpDir.removeCallback()

  return messages
}

export default extractMessages
