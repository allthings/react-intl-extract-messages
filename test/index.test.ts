import { join, resolve } from 'path'
import { exec } from 'child_process'
import { extractMessages } from '../src'

const tsconfig = join(__dirname, 'fixture-project/tsconfig.json')
const badTsconfig = join(__dirname, 'fixture-project/bad-tsconfig.json')

jest.setTimeout(10000)

test('Extract all messages', () => {
  const messages = extractMessages(tsconfig)

  expect(messages).toMatchSnapshot()
})

test('Extract all messages with options', () => {
  const messages = extractMessages(tsconfig, {
    additionalComponentNames: {
      'react-lingo': ['default', 'LingoFormattedMessage'],
    },
  })

  expect(messages).toMatchSnapshot()
})

test('Fail with invalid non existent tsconfig file', () => {
  expect(() => {
    extractMessages('invalid file')
  }).toThrow()
})

test('Fail with bad tsconfig file', () => {
  expect(() => {
    extractMessages(badTsconfig)
  }).toThrow()
})

test('Test binary', done => {
  exec(
    'node_modules/.bin/ts-node src/cli.ts --tsconfig test/fixture-project/tsconfig.json --options-file test/fixture-project/opts.json',
    { cwd: resolve(__dirname, '..') },
    (err, stdout) => {
      expect(err).toBe(null)
      expect(JSON.parse(stdout)).toMatchSnapshot()
      done()
    },
  )
})
