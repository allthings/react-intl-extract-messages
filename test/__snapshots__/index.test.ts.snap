// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Extract all messages 1`] = `
Array [
  Object {
    "defaultMessage": "Privacy",
    "description": "privacy",
    "id": "footer.privacy",
  },
  Object {
    "defaultMessage": "Terms",
    "description": "terms",
    "id": "footer.terms",
  },
  Object {
    "defaultMessage": "Settings",
    "description": "settings",
    "id": "footer.settings",
  },
  Object {
    "defaultMessage": "Hello {name}",
    "description": "hello",
    "id": "header.hello",
  },
  Object {
    "defaultMessage": "<a href=\\"#\\">You can change your cookie settings here</a>",
    "description": "cookie",
    "id": "footer.cookie",
  },
  Object {
    "defaultMessage": "external",
    "description": "external",
    "id": "external.external",
  },
]
`;

exports[`Extract all messages with options 1`] = `
Array [
  Object {
    "defaultMessage": "Privacy",
    "description": "privacy",
    "id": "footer.privacy",
  },
  Object {
    "defaultMessage": "Terms",
    "description": "terms",
    "id": "footer.terms",
  },
  Object {
    "defaultMessage": "Settings",
    "description": "settings",
    "id": "footer.settings",
  },
  Object {
    "defaultMessage": "Hello {name}",
    "description": "hello",
    "id": "header.hello",
  },
  Object {
    "defaultMessage": "<a href=\\"#\\">You can change your cookie settings here</a>",
    "description": "cookie",
    "id": "footer.cookie",
  },
  Object {
    "defaultMessage": "foo",
    "description": "foo",
    "id": "foo",
  },
  Object {
    "defaultMessage": "bar",
    "description": "bar",
    "id": "bar",
  },
  Object {
    "defaultMessage": "external",
    "description": "external",
    "id": "external.external",
  },
]
`;

exports[`Test binary 1`] = `
Array [
  Object {
    "defaultMessage": "Privacy",
    "description": "privacy",
    "id": "footer.privacy",
  },
  Object {
    "defaultMessage": "Terms",
    "description": "terms",
    "id": "footer.terms",
  },
  Object {
    "defaultMessage": "Settings",
    "description": "settings",
    "id": "footer.settings",
  },
  Object {
    "defaultMessage": "Hello {name}",
    "description": "hello",
    "id": "header.hello",
  },
  Object {
    "defaultMessage": "<a href=\\"#\\">You can change your cookie settings here</a>",
    "description": "cookie",
    "id": "footer.cookie",
  },
  Object {
    "defaultMessage": "foo",
    "description": "foo",
    "id": "foo",
  },
  Object {
    "defaultMessage": "bar",
    "description": "bar",
    "id": "bar",
  },
  Object {
    "defaultMessage": "external",
    "description": "external",
    "id": "external.external",
  },
]
`;
